File: haskell-migration/ebm-mars/src/EBM/PhaseChange.hs
{-|
Module      : EBM.PhaseChange
Description : Phase change processes (CO2 ice/regolith dynamics)
Copyright   : (c) 2026
License     : MIT

Functions for calculating CO2 phase changes, including sublimation,
ice formation, and regolith adsorption on Mars.
-}

module EBM.PhaseChange
  ( SublimationParams(..)
  , calcSublimationParams
  , calcTemperatureMass
  , calcSlopeTemperatureMass
  , calcIcePressure
  , calcRegolithPressure
  ) where

import EBM.Types
import qualified Data.Vector.Unboxed as V

-- Physical constants
latentHeat :: Double
latentHeat = 5.9e5  -- Latent heat of CO2 [J/kg]

heatCapacity :: Double
heatCapacity = 1.0e7  -- Heat capacity [J/K/m^2]

iceMassFactor :: Double
iceMassFactor = 5.815e-6  -- Conversion factor [bar/(kg/m^2)]

-- | Sublimation parameters (temperature and albedos)
data SublimationParams = SublimationParams
  { sublimationTemperature :: !Double  -- [K]
  , albedoFrost :: !Double             -- Frost albedo
  , albedoIce :: !Double               -- Ice albedo
  } deriving (Show, Eq)

-- | Calculate sublimation parameters from pressure
calcSublimationParams :: Pressure -> SublimationParams
calcSublimationParams (Pressure pAir) =
  let -- Polynomial coefficients for sublimation temperature
      tCoeffs = [194.36, 26.451, 2.8593, 0.1814, 0.0046]
      
      -- Polynomial coefficients for frost albedo
      afCoeffs = [0.21, -0.0008, -0.0074, -0.0147, 0.0337, 0.1381, 0.3249]
      
      -- Polynomial coefficients for ice albedo
      aiCoeffs = [0.63, -0.0008, -0.0011, 0.0183, 0.0599, 0.6997]
      
      x = if pAir >= 1e-16
          then log pAir / log 10.0  -- log10
          else -16.0
      
      -- Calculate sublimation temperature
      tsub = sum $ zipWith (\c i -> c * x ** fromIntegral i) tCoeffs [0..]
      
      -- Calculate albedos
      (af, ai) = if pAir >= 1e-3
                 then let af' = afCoeffs !! 1 * x**5 + afCoeffs !! 2 * x**4 + 
                               afCoeffs !! 3 * x**3 + afCoeffs !! 4 * x**2 + 
                               afCoeffs !! 5 * x + afCoeffs !! 6
                          ai' = aiCoeffs !! 1 * x**4 + aiCoeffs !! 2 * x**3 + 
                               aiCoeffs !! 3 * x**2 + aiCoeffs !! 4 * x + 
                               aiCoeffs !! 5 - 0.08
                      in (af', ai')
                 else (afCoeffs !! 0, aiCoeffs !! 0 - 0.08)
      
  in SublimationParams tsub af ai

-- | Calculate temperature and mass evolution (global)
calcTemperatureMass :: SimulationConfig -> PlanetParams -> ClimateState 
                    -> V.Vector Double -> V.Vector Double -> V.Vector Double
                    -> ClimateState
calcTemperatureMass config params state ins dif radi =
  let reso = resolution config
      dt = unTime $ timeStep config
      l = latentHeat
      c = heatCapacity
      
      subParams = calcSublimationParams (airPressure state)
      tsub = sublimationTemperature subParams
      af = albedoFrost subParams
      ai = albedoIce subParams
      
      temps = temperatures state
      masses = co2Masses state
      
      -- Update temperature and mass for each latitude
      (newTemps, newMasses) = V.unzip $ V.generate reso $ \n ->
        let t = temps V.! n
            m = masses V.! n
            insN = ins V.! n
            difN = dif V.! n
            radiN = radi V.! n
        in if m == 0.0
           then -- No ice present
             let delE = (insN * (1.0 - af) + difN - radiN) * dt
                 tNew = t + delE / c
             in if tNew < tsub
                then (tsub, (tsub - tNew) * c / l)  -- Ice forms
                else (tNew, 0.0)
           else -- Ice present
             let delE = (insN * (1.0 - ai) + difN - radiN) * dt
                 mNew = m - delE / l + (tsub - t) * c / l
             in if mNew < 0.0
                then (tsub + (-mNew) * l / c, 0.0)  -- Ice sublimates
                else (tsub, mNew)
      
  in state { temperatures = newTemps
           , co2Masses = newMasses
           , sublimationTemp = Temperature tsub
           }

-- | Calculate temperature and mass evolution (slope)
calcSlopeTemperatureMass :: SimulationConfig -> PlanetParams -> ClimateState
                         -> V.Vector Double -> V.Vector Double  -- Current T and M
                         -> V.Vector Double -> V.Vector Double -> V.Vector Double  -- ins, radi, dif
                         -> (V.Vector Double, V.Vector Double)  -- New T and M
calcSlopeTemperatureMass config params state slopeT slopeM ins radi dif =
  let reso = resolution config
      dt = unTime $ timeStep config
      l = latentHeat
      c = heatCapacity
      
      subParams = calcSublimationParams (airPressure state)
      tsub = sublimationTemperature subParams
      af = albedoFrost subParams
      ai = albedoIce subParams
      
      -- Update temperature and mass for each latitude
      V.unzip $ V.generate reso $ \n ->
        let t = slopeT V.! n
            m = slopeM V.! n
            insN = ins V.! n
            difN = dif V.! n
            radiN = radi V.! n
        in if m == 0.0
           then -- No ice present
             let delE = (insN * (1.0 - af) + difN - radiN) * dt
                 tNew = t + delE / c
             in if tNew < tsub
                then (tsub, (tsub - tNew) * c / l)  -- Ice forms
                else (tNew, 0.0)
           else -- Ice present
             let delE = (insN * (1.0 - ai) + difN - radiN) * dt
                 mNew = m - delE / l + (tsub - t) * c / l
             in if mNew < 0.0
                then (tsub + (-mNew) * l / c, 0.0)  -- Ice sublimates
                else (tsub, mNew)

-- | Calculate ice pressure from CO2 mass distribution
calcIcePressure :: SimulationConfig -> PlanetParams -> ClimateState -> Pressure
calcIcePressure config params state =
  let reso = resolution config
      masses = co2Masses state
      factor = iceMassFactor
      
      -- Integrate over latitude bands
      pice = V.sum $ V.generate reso $ \n ->
        let upper = 90.0 / 181.0 * (2.0 * fromIntegral n - 179.0) * pi / 180.0
            lower = 90.0 / 181.0 * (2.0 * fromIntegral n - 181.0) * pi / 180.0
            scale = 2.0 * pi * (sin upper - sin lower)
        in masses V.! n * factor * scale
      
  in if pice > unPressure (totalPressure params)
     then totalPressure params
     else Pressure pice

-- | Calculate regolith pressure (adsorbed CO2)
calcRegolithPressure :: SimulationConfig -> PlanetParams -> ClimateState -> (Pressure, Pressure, Double)
calcRegolithPressure config params state =
  let reso = resolution config
      cConst = 34.0
      tD = 35.0
      gamma = 0.275
      
      temps = temperatures state
      tsub = unTemperature $ sublimationTemp state
      pIce = unPressure $ icePressure state
      pTotal = unPressure $ totalPressure params
      
      -- Calculate sigma (regolith capacity)
      sigma = V.sum $ V.generate reso $ \n ->
        let theta = fromIntegral (n - 90) * pi / 180.0
            t = temps V.! n
        in if t > tsub
           then cConst * exp (-t / tD) * cos theta * (pi / 180.0)
           else 0.0
      
      -- Solve for atmospheric pressure using bisection
      (pAir, pRego, bug) = solvePressure sigma pIce pTotal gamma 0 100
      
  in (Pressure pAir, Pressure pRego, bug)
  where
    -- Bisection solver for pressure balance equation
    solvePressure :: Double -> Double -> Double -> Double -> Int -> Int -> (Double, Double, Double)
    solvePressure sigma pIce pTotal gamma loop loopMax =
      let kouho0 = pTotal / 2.0
          limHi0 = pTotal
          limLo0 = 0.0
      in go kouho0 limHi0 limLo0 loop
      where
        go kouho limHi limLo loopCount
          | loopCount >= loopMax = (kouho, sigma * kouho ** gamma, 0.0)
          | otherwise =
              let fx = sigma * kouho ** gamma + kouho + pIce - pTotal
              in if fx == 0.0
                 then (kouho, sigma * kouho ** gamma, if kouho < 0 then 1.0 else 0.0)
                 else if fx > 0.0
                      then let kouhoNew = 0.5 * (kouho + limLo)
                           in go kouhoNew kouho limLo (loopCount + 1)
                      else let kouhoNew = 0.5 * (limHi + kouho)
                           in go kouhoNew limHi kouho (loopCount + 1)

File: ebm-c-refactored/phase_change.c
#include "ebm.h"

void Tsub_af_ai(double P_air, double *Tsub, double *a_f, double *a_i) {
  double t[5], af[7], ai[6];
  double x;
  t[0] = 194.36;
  t[1] = 26.451;
  t[2] = 2.8593;
  t[3] = 0.1814;
  t[4] = 0.0046;
  af[0] = 0.21;
  af[1] = -0.0008;
  af[2] = -0.0074;
  af[3] = -0.0147;
  af[4] = 0.0337;
  af[5] = 0.1381;
  af[6] = 0.3249;
  ai[0] = 0.63;
  ai[1] = -0.0008;
  ai[2] = -0.0011;
  ai[3] = 0.0183;
  ai[4] = 0.0599;
  ai[5] = 0.6997;
  x = log10(P_air);

  if (P_air >= 1e-3) {
    *a_f = af[1] * pow(x, 5.0) + af[2] * pow(x, 4.0) + af[3] * pow(x, 3.0) +
           af[4] * pow(x, 2.0) + af[5] * x + af[6];
    *a_i = ai[1] * pow(x, 4.0) + ai[2] * pow(x, 3.0) + ai[3] * pow(x, 2.0) +
           ai[4] * x + ai[5] - 0.08;
  } else {
    *a_f = af[0];
    *a_i = ai[0] - 0.08;
  }

  if (P_air >= 1e-16) {
    *Tsub = t[0] + t[1] * x + t[2] * pow(x, 2.0) + t[3] * pow(x, 3.0) +
            t[4] * pow(x, 4.0);
  } else {
    x = -16.0;
    *Tsub = t[0] + t[1] * x + t[2] * pow(x, 2.0) + t[3] * pow(x, 3.0) +
            t[4] * pow(x, 4.0);
  }
}

void calc_T_M(const SimulationConfig *conf, const PlanetParams *para,
              ClimateState *state, double ins[], double dif[], double radi[]) {
  int reso = conf->reso;
  int n;
  double Tsub = 0.0, a_i = 0.0, a_f = 0.0, L = 5.9e5, C = 1.0e7;
  double del_E[reso];

  Tsub_af_ai(state->P_air, &Tsub, &a_f, &a_i);
  state->T_sub = Tsub;

  for (n = 0; n < reso; n++) {
    if (state->M[n] == 0.0) {
      del_E[n] = (ins[n] * (1.0 - a_f) + dif[n] - radi[n]) * conf->dt;
      state->T[n] = state->T[n] + del_E[n] / C;
      state->M[n] = 0.0;
      if (state->T[n] < Tsub) {
        state->M[n] = (Tsub - state->T[n]) * C / L;
        state->T[n] = Tsub;
      }
    } else if (state->M[n] != 0.0) {
      del_E[n] = (ins[n] * (1.0 - a_i) + dif[n] - radi[n]) * conf->dt;
      state->M[n] = state->M[n] - del_E[n] / L + (Tsub - state->T[n]) * C / L;
      state->T[n] = Tsub;
      if (state->M[n] < 0.0) {
        state->T[n] = Tsub + (-state->M[n]) * L / C;
        state->M[n] = 0.0;
      }
    }
  }
}

void calc_ice(const SimulationConfig *conf, const PlanetParams *para,
              ClimateState *state) {
  int reso = conf->reso;
  int n;
  double upper = 0.0, lower = 0.0, scale = 0.0;
  double Pice = 0.0, factor = 5.815e-6;

  for (n = 0; n < reso; n++) {
    upper = 90.0 / 181.0 * (2.0 * n - 179.0) * M_PI / 180.0;
    lower = 90.0 / 181.0 * (2.0 * n - 181.0) * M_PI / 180.0;
    scale = 2.0 * M_PI * (sin(upper) - sin(lower));
    Pice += state->M[n] * factor * scale;
  }
  if (Pice > para->P_total)
    state->P_ice = para->P_total;
  else
    state->P_ice = Pice;
}

void calc_rego(const SimulationConfig *conf, const PlanetParams *para,
               ClimateState *state) {
  int reso = conf->reso;
  int n, loop = 0, loop_max = 100;
  double C = 34.0, T_d = 35.0, gamma = 0.275;
  double sigma = 0.0;
  double kouho = para->P_total / 2.0;
  double theta[reso];
  double lim_hi = para->P_total, lim_lo = 0.0, fx = 0.0;

  for (n = 0; n < reso; n++) {
    theta[n] = (n - 90) * M_PI / 180;
    if (state->T[n] > state->T_sub) {
      sigma =
          sigma + C * exp(-state->T[n] / T_d) * cos(theta[n]) * (M_PI / 180.0);
    }
  }

  do {
    fx = sigma * pow(kouho, gamma) + kouho + state->P_ice - para->P_total;
    loop = loop + 1;
    if (fx > 0.0) {
      lim_hi = kouho;
      kouho = 0.5 * (lim_hi + lim_lo);
    } else if (fx < 0.0) {
      lim_lo = kouho;
      kouho = 0.5 * (lim_hi + lim_lo);
    }
  } while (fx != 0.0 && loop < loop_max);

  state->P_air = kouho;
  state->P_rego = sigma * pow(kouho, gamma);

  if (kouho < 0) {
    state->bug = 1;
  } else {
    state->bug = 0;
  }
}

void calc_Ts_Ms(const SimulationConfig *conf, const PlanetParams *para,
                ClimateState *state, double *T, double *M, double ins[],
                double radi[], double dif[]) {
  int reso = conf->reso;
  int n;
  double a_i = 0.63, a_f = 0.21, L = 5.9e5, C = 1.0e7, Tsub = 0.0;
  double del_E[reso];

  Tsub_af_ai(state->P_air, &Tsub, &a_f, &a_i);

  for (n = 0; n < reso; n++) {
    if (M[n] == 0.0) {
      del_E[n] = (ins[n] * (1.0 - a_f) + dif[n] - radi[n]) * conf->dt;
      T[n] = T[n] + del_E[n] / C;
      M[n] = 0.0;
      if (T[n] < Tsub) {
        M[n] = (Tsub - T[n]) * C / L;
        T[n] = Tsub;
      }
    } else if (M[n] != 0.0) {
      del_E[n] = (ins[n] * (1.0 - a_i) + dif[n] - radi[n]) * conf->dt;
      M[n] = M[n] - del_E[n] / L + (Tsub - T[n]) * C / L;
      T[n] = Tsub;
      if (M[n] < 0.0) {
        T[n] = Tsub + (-M[n]) * L / C;
        M[n] = 0.0;
      }
    }
  }
}
